README.md
# LandBlock_for-1
Seablock but on land


LandBlock_for_1_1.1.1/changelog.txt
 ---------------------------------------------------------------------------------------------------
Version: 1.0.1
Date: 2020.12.16
  Bugs:
     - Fixed no file at icon path for sb-blue-algae-liquefaction
  Changes:
     - Threre is now a Changelog
---------------------------------------------------------------------------------------------------
Version: 1.0.0
Date: 2020.11.01
  Changes: 
     - Now supports version 1.0.0
     - Has all Seablock Balance changes from Seablock 4.8
---------------------------------------------------------------------------------------------------

LandBlock_for_1_1.1.1/data.lua
angelsmods = angelsmods or {}

require "prototypes/recipe"
require "prototypes/recipe-category"
require "prototypes/technology"
require "prototypes/rockchest"

local lib = require "lib"

table.insert(data.raw.character.character.crafting_categories, "crafting-handonly")

-- for _,v in pairs(data.raw.tile) do
--   v.autoplace = nil
-- end

-- Want player to collide with cliffs
-- Want player to collide with water
-- Don't want cliffs to collide with water
-- Water tile's default collision mask includes player-layer and item-layer
-- So use train-layer as a substitute player-layer
-- data.raw.cliff['cliff'].collision_mask =
--   {"object-layer", "train-layer", "not-colliding-with-itself"}

-- local octaves = -3
-- local persistence = 0.2
-- local waterline = 9.4
-- local elevation_scale = 5
-- local function scale_elevation(x)
--   return (x - waterline) * elevation_scale + waterline
-- end
-- low lying sand
-- data.raw.tile['sand-1'].autoplace = {
--   peaks = {
--   { -- Around cliff islands
--     influence = 5,
--     elevation_optimal = 0.3 * elevation_scale + waterline,
--     elevation_range = 0.3 * elevation_scale,
--     elevation_max_range = 0.3 * elevation_scale
--   },
--   {
--     influence = 0.77 * 8, -- Worm islands
--     min_influence = 0,
--     noise_layer = 'enemy-base',
--     noise_octaves_difference = octaves,
--     noise_persistence = persistence,
--     tier_from_start_optimal = 8,
--     tier_from_start_max_range = 40,
--     tier_from_start_top_property_limit = 8
--   },
--   { -- Not in starting area
--     influence = -5,
--     starting_area_weight_optimal = 1,
--     starting_area_weight_range = 0,
--     starting_area_weight_max_range = 0
--   },
--   {
--     influence = 100, -- ... except for starting tile
--     min_influence = 0,
--     distance_optimal = 0,
--     distance_range = 0.1,
--     distance_max_range = 0.1
--   },
--   {
--     influence = -5
--   }
-- }}

-- highground sand
-- data.raw.tile['sand-2'].autoplace = {
--   peaks = {
--   {
--     influence = 5,
--     min_influence = 0,
--     elevation_optimal = scale_elevation(15),
--     elevation_range = 5 * elevation_scale,
--     elevation_max_range = 5 * elevation_scale,
--     tier_from_start_optimal = 0.1,
--     tier_from_start_range = 0,
--     tier_from_start_max_range = 0,
--     tier_from_start_top_property_limit = 0.1
--   },
--   {
--     influence = 0.65, -- starting area garden islands
--     min_influence = 0,
--     max_influence = 1,
--     noise_layer = 'enemy-base',
--     noise_octaves_difference = octaves,
--     noise_persistence = persistence,
--     tier_from_start_optimal = 0,
--     tier_from_start_range = 0.1,
--     tier_from_start_max_range = 0.1
--   },
--   {
--     influence = 1,
--     max_influence = 0,
--     starting_area_weight_optimal = 1,
--     starting_area_weight_range = 0,
--     starting_area_weight_max_range = 0
--   },
--   {
--     influence = -100, -- not on starting tile
--     max_influence = 0,
--     distance_optimal = 0,
--     distance_range = 3,
--     distance_max_range = 3
--   },
-- }}

-- local plant_elevation_range = 9.9 * elevation_scale
-- data.raw.tree['desert-garden'].autoplace = {
--   max_probability = 0.2,
--   random_probability_penalty = 0.05,
--   sharpness = 1,
--   peaks = {
--     {
--       influence = 1,
--       min_influence = 0,
--       elevation_optimal = scale_elevation(20),
--       elevation_range = plant_elevation_range,
--       elevation_max_range = plant_elevation_range
--     },
--     {
--       influence = 0.31, -- Trial and error value to generate size that approximately matches starting area islands
--       min_influence = 0,
--       max_influence = 1,
--       noise_layer = 'enemy-base',
--       noise_octaves_difference = octaves,
--       noise_persistence = persistence,
--       tier_from_start_optimal = 0,
--       tier_from_start_range = 0.1,
--       tier_from_start_max_range = 0.1
--     },
--     {
--       influence = 1,
--       max_influence = 0,
--       noise_layer = 'desert-garden-noise',
--       noise_persistence = 0.8,
--       noise_octaves_difference = -0.5
--     }
--   },
--   order = "yc",
--   tile_restriction = {'sand-2'}
-- }

-- data.raw.tree['temperate-garden'].autoplace = {
--   max_probability = 0.2,
--   random_probability_penalty = 0.05,
--   peaks = {
--     {
--       influence = 1,
--       min_influence = 0,
--       elevation_optimal = scale_elevation(20),
--       elevation_range = plant_elevation_range,
--       elevation_max_range = plant_elevation_range
--     },
--     {
--       influence = 1,
--       max_influence = 0,
--       noise_layer = 'temperate-garden-noise',
--       noise_persistence = 0.8,
--       noise_octaves_difference = -0.5
--     }
--   },
--   order = "ya",
--   tile_restriction = {'sand-2'}
-- }

-- data.raw.tree['swamp-garden'].autoplace = {
--   max_probability = 0.05,
--   peaks = {
--     {
--       influence = 1,
--       min_influence = 0,
--       elevation_optimal = scale_elevation(20),
--       elevation_range = plant_elevation_range,
--       elevation_max_range = plant_elevation_range
--     }
--   },
--   order = "yb",
--   tile_restriction = {'sand-2'}
-- }

-- data.raw.tree['desert-tree'].autoplace = {
--   max_probability = 0.1,
--   random_probability_penalty = 0.025,
--   sharpness = 1,
--   peaks = {
--     {
--       influence = 1,
--       min_influence = 0,
--       elevation_optimal = scale_elevation(20),
--       elevation_range = plant_elevation_range,
--       elevation_max_range = plant_elevation_range
--     },
--     {
--       influence = 0.31, -- Trial and error value to generate size that approximately matches starting area islands
--       min_influence = 0,
--       max_influence = 1,
--       noise_layer = 'enemy-base',
--       noise_octaves_difference = octaves,
--       noise_persistence = persistence,
--       tier_from_start_optimal = 0,
--       tier_from_start_range = 0.1,
--       tier_from_start_max_range = 0.1
--     },
--     {
--       influence = 1,
--       max_influence = 0,
--       noise_layer = 'desert-tree-noise',
--       noise_persistence = 0.8,
--       noise_octaves_difference = -0.5
--     }
--   },
--   order = "za",
--   tile_restriction = {'sand-2'}
-- }
-- data.raw.tree['temperate-tree'].autoplace = {
--   max_probability = 0.1,
--   random_probability_penalty = 0.025,
--   peaks = {
--     {
--       influence = 1,
--       min_influence = 0,
--       elevation_optimal = scale_elevation(20),
--       elevation_range = plant_elevation_range,
--       elevation_max_range = plant_elevation_range
--     },
--     {
--       influence = 1,
--       max_influence = 0,
--       noise_layer = 'temperate-tree-noise',
--       noise_persistence = 0.8,
--       noise_octaves_difference = -0.5
--     }
--   },
--   order = "zc",
--   tile_restriction = {'sand-2'}
-- }

-- data.raw.tree['swamp-tree'].autoplace = {
--   max_probability = 0.05,
--   peaks = {
--     {
--       influence = 1,
--       min_influence = 0,
--       elevation_optimal = scale_elevation(20),
--       elevation_range = plant_elevation_range,
--       elevation_max_range = plant_elevation_range
--     },
--     {
--       influence = 1,
--       max_influence = 0,
--       noise_layer = 'swamp-tree-noise',
--       noise_persistence = 0.8,
--       noise_octaves_difference = -0.5
--     }
--   },
--   order = "zb",
--   tile_restriction = {'sand-2'}
-- }

-- data.raw.tile['water'].autoplace = {
--   peaks = {
--   {
--     influence = 0.1, -- shallow water around cliff islands
--     min_influence = 0,
--     elevation_optimal = -2 * elevation_scale + waterline,
--     elevation_range = 2.5 * elevation_scale,
--     elevation_max_range = 2.5 * elevation_scale
--   },
--   {
--     influence = 0.77 * 2, -- around worm islands
--     min_influence = 0,
--     max_influence = 1,
--     noise_layer = 'enemy-base',
--     noise_octaves_difference = octaves,
--     noise_persistence = persistence,
--   },
--   {
--     influence = 5, -- around starting tile
--     min_influence = 0,
--     distance_optimal = 0,
--     distance_range = 5,
--     distance_max_range = 5
--   }
-- }}

-- data.raw.tile['deepwater'].autoplace = {
--   peaks = {
--   {
--     influence = 0.01
--   }
-- }}

-- data.raw.fish['alien-fish-1'].autoplace = {
--   peaks = {
--   {
--     influence = -0.1,
--     max_influence = 0,
--     noise_layer = 'enemy-base',
--     noise_octaves_difference = octaves,
--     noise_persistence = persistence,
--   },
--   {
--     influence = 0.01
--   }}
-- }

-- data.raw.fish['alien-fish-2'].autoplace = {
--   peaks = {
--   {
--     influence = 0.02,
--     min_influence = 0,
--     noise_layer = 'enemy-base',
--     noise_octaves_difference = octaves,
--     noise_persistence = persistence,
--   }}
-- }

-- data.raw.fish['alien-fish-3'].autoplace = {
--   peaks = {
--   {
--     influence = 0.015,
--     min_influence = 0,
--     elevation_optimal = scale_elevation(10),
--     elevation_range = 5 * elevation_scale,
--     elevation_max_range = 5 * elevation_scale,
--     tier_from_start_optimal = 0.1,
--     tier_from_start_range = 0,
--     tier_from_start_max_range = 0,
--     tier_from_start_top_property_limit = 0.1
--   }}
-- }
--[[
local function worm_autoplace(distance, probability, order, falloff)
  local ret = {
    sharpness = 1,
    control = "enemy-base",
    richness_multiplier = 1,
    richness_base = 0,
    force = "enemy",
    max_probability = probability or 0.01,
    random_probability_penalty = 0.05,
    order = order or 'a',
    peaks = {
      {
        influence = 10,
	richness_influence = 100,
	tier_from_start_optimal = distance,
	tier_from_start_top_property_limit = distance + falloff * 2,
	tier_from_start_max_range = 2,
        noise_layer = 'enemy-base',
        noise_octaves_difference = octaves,
        noise_persistence = persistence
      },
      {
        influence = 10,
	elevation_optimal = scale_elevation(4),
	elevation_range = 5 * elevation_scale,
	elevation_max_range = 5 * elevation_scale
      },
      {
        influence = -10,
        starting_area_weight_optimal = 1,
        starting_area_weight_range = 0.033,
        starting_area_weight_max_range = 0.033
      },
      {
        influence = -21
      }
    }
  }
  return ret
end
]]--
-- local noise = require "noise"
-- local tne = noise.to_noise_expression
-- local enemy_random_seed = 1
-- local function new_random_seed()
--   enemy_random_seed = enemy_random_seed + 1
--   return enemy_random_seed
-- end
-- local function worm_autoplace(distance, probability, order, falloff)


--     -- no worms
--     probability = 0

--   local d = noise.var("distance") - noise.var("starting_area_radius")
--   p = noise.clamp((d - distance * 128) / 128, 0, 1)
--   if falloff then
--     p = p * noise.clamp(((distance + 2) * 128 - d) / 128, 0, 1)
--   end
--   p = p * noise.clamp((waterline - noise.var("elevation")), 0, 1)
--   p = p * probability
--   p = noise.random_penalty(p, probability * 0.5, {
--     x = noise.var("x") + new_random_seed()
--   })

--   return {
--     order = order,
--     force = "enemy",
--     probability_expression = p,
--     richness_expression = tne(1)
--   }
-- end

-- data.raw.turret['small-worm-turret'].autoplace = worm_autoplace(0, 1, 'z', true)
-- data.raw.turret['medium-worm-turret'].autoplace = worm_autoplace(1, 1, 'y', true)
-- if data.raw.turret['bob-big-explosive-worm-turret'] then
--   data.raw.turret['bob-big-explosive-worm-turret'].autoplace = worm_autoplace(1.5, 0.5, 'v', false)
-- end
-- if data.raw.turret['bob-big-fire-worm-turret'] then
--   data.raw.turret['bob-big-fire-worm-turret'].autoplace = worm_autoplace(1.5, 0.5, 'v', false)
-- end
-- if data.raw.turret['bob-big-poison-worm-turret'] then
--   data.raw.turret['bob-big-poison-worm-turret'].autoplace = worm_autoplace(1.5, 0.5, 'v', false)
-- end
-- if data.raw.turret['bob-big-piercing-worm-turret'] then
--   data.raw.turret['bob-big-piercing-worm-turret'].autoplace = worm_autoplace(1.5, 0.5, 'v', false)
-- end
-- if data.raw.turret['bob-big-electric-worm-turret'] then
--   data.raw.turret['bob-big-electric-worm-turret'].autoplace = worm_autoplace(1.5, 0.5, 'v', false)
-- end
-- if data.raw.turret['bob-giant-worm-turret'] then
--   data.raw.turret['bob-giant-worm-turret'].autoplace = worm_autoplace(2, 0.6, 'u', false)
-- end
-- if data.raw.turret['behemoth-worm-turret'] then
--   data.raw.turret['big-worm-turret'].autoplace = worm_autoplace(1.5, 0.5, 'v', false)
--   data.raw.turret['behemoth-worm-turret'].autoplace = worm_autoplace(2, 0.2, 't', false)
-- else
--   data.raw.turret['big-worm-turret'].autoplace = worm_autoplace(2, 1, 'v', false)
-- end
-- data.raw.tree['puffer-nest'].autoplace = worm_autoplace(0, 0.01, 's', false)

-- for _,v in pairs(data.raw.turret) do
--   v.map_generator_bounding_box = nil
-- end

-- data:extend({
--   {
--     type = 'noise-layer',
--     name = 'desert-tree-noise'
--   },
--   {
--     type = 'noise-layer',
--     name = 'temperate-tree-noise'
--   },
--   {
--     type = 'noise-layer',
--     name = 'swamp-tree-noise'
--   },
--   {
--     type = 'noise-layer',
--     name = 'desert-garden-noise'
--   },
--   {
--     type = 'noise-layer',
--     name = 'temperate-garden-noise'
--   },
--   {
--     type = 'noise-layer',
--     name = 'swamp-garden-noise'
--   }
-- })
--[[
autoplace = data.raw.tile['water'].autoplace
for _,peak in pairs(autoplace.peaks) do
  print("")
  for k,v in pairs(peak) do
    print(k .. " = " .. v)
  end
end
]]--
--[[
local function search(t, path, s)
  for k,v in pairs(t) do
    p = path .. "." .. k
    if string.find(k, s) then
      print(p)
    end
    if type(v) == 'string' and string.find(v, s) then
      print(p .. '=' .. v)
    end
    if type(v) == 'table' then
      search(v, p, s)
    end
  end
end
search(data.raw, "data.raw", "autoplace")
]]--

-- local function make_basis_noise_function(seed0,seed1,outscale0,inscale0)
--   outscale0 = outscale0 or 1
--   inscale0 = inscale0 or 1/outscale0
--   return function(x,y,inscale,outscale)
--     return tne{
--       type = "function-application",
--       function_name = "factorio-basis-noise",
--       arguments = {
--         x = tne(x),
--         y = tne(y),
--         seed0 = tne(seed0),
--         seed1 = tne(seed1),
--         input_scale = tne((inscale or 1) * inscale0),
--         output_scale = tne((outscale or 1) * outscale0)
--       }
--     }
--   end
-- end

-- data.raw['noise-expression']['cliffiness'].expression =
--   noise.define_noise_function(function(x, y, tile, map)
--     local t = noise.clamp(tile.tier - 0.2, 0, 1) -- No cliffs in starting area
--     return 100 * t
--   end)
-- data.raw['noise-expression']['elevation'].expression =
--   noise.define_noise_function(function(x, y, tile, map)
--       x = x + 40000
--       y = y
--       local v = make_basis_noise_function(map.seed, 5, 6, 1/64)(x, y)
--       v = noise.max(v, 0)
--       v = (v * elevation_scale) - (waterline * (elevation_scale - 1)) -- Increase gradient for cliffs while leaving waterline unchanged
--       return v
--   end)


LandBlock_for_1_1.1.1/control.lua
landblock = {}

landblock.giveresearch = function(force)
  local techs = {
    'landfill',
    'bio-paper-1',
    'bio-processing-brown'
  }
  local newforce = force.technologies['sb-startup1'].researched == false
  for _,v in ipairs(techs) do
    force.technologies[v].researched = true
  end
  if newforce then
    force.add_research("sb-startup1")
  end
end

landblock.giveitems = function(entity)
  local landfill = 'landfill'
  if game.item_prototypes['landfill-sand-3'] then
    landfill = 'landfill-sand-3'
  end
  local stuff = {
    {landfill, 1000},
    {"stone", 50},
    {"small-electric-pole", 50},
    {"small-lamp", 12},

    {"iron-plate", 1200},
    {"basic-circuit-board", 200},
    {"stone-pipe", 100},
    {"stone-pipe-to-ground", 50},
    {"stone-brick", 500},
    {"pipe", 27},
    {"copper-pipe", 5},
    {"iron-gear-wheel", 25},
    {"iron-stick", 96},
    {"pipe-to-ground", 2},
    {"electronic-circuit", 10}
  }
  if game.item_prototypes["wind-turbine-2"] then
    table.insert(stuff, {"wind-turbine-2", 120})
  else
    table.insert(stuff, {"solar-panel", 38})
    table.insert(stuff, {"accumulator", 32})
  end
  for _,v in ipairs(stuff) do
    entity.insert{name = v[1], count = v[2]}
  end
end

script.on_event(defines.events.on_player_joined_game, function(e)
  landblock.giveresearch(game.players[e.player_index].force)
end)

script.on_event(defines.events.on_force_created, function(e)
  landblock.giveresearch(e.force)
end)

script.on_event(defines.events.on_player_created, function(e)
  local player = game.players[e.player_index]
  player.remove_item{name = 'burner-mining-drill', count = 1}
  player.remove_item{name = 'wood', count = 1}
end)

script.on_event(defines.events.on_chunk_generated, function(e)
  local surface = e.surface;
  if surface.name ~= "nauvis" and surface.name:sub(1, 14) ~= "battle_surface" then
    return
  end
  local ltx = e.area.left_top.x
  local lty = e.area.left_top.y
  local rbx = e.area.right_bottom.x
  local rby = e.area.right_bottom.y
  for k,v in pairs(surface.map_gen_settings.starting_points) do
    if v.x >= ltx and v.y >= lty and v.x < rbx and v.y < rby then
      local chest = surface.create_entity({name = "rock-chest", position = v, force = game.forces.neutral})
      landblock.giveitems(chest)
    end
  end
end)

local function setpvp()
  if remote.interfaces.pvp then
    remote.call("pvp", "set_config", {silo_offset = {x = 16, y = 16}})
  end
end

local function init()
  setpvp()
  if remote.interfaces.freeplay and remote.interfaces.freeplay.set_disable_crashsite then
    remote.call("freeplay", "set_disable_crashsite", true)
  end
  global.unlocks = {
    ['angels-ore3-crushed'] = 'sb-startup1',
    ['basic-circuit-board'] = 'sb-startup2',
    ['algae-green'] = 'bio-wood-processing',
  }
  if game.technology_prototypes['sct-automation-science-pack'] then
    global.unlocks['lab'] = 'sct-automation-science-pack'
  else
    global.unlocks['lab'] = 'sb-startup4'
  end
end
local function haveitem(player, itemname, crafted)
  local unlock = global.unlocks[itemname]
  -- Special case for basic-circuit because it is part of starting equipment
  if unlock and (itemname ~= 'basic-circuit-board' or crafted) then
    player.force.technologies[unlock].researched = true
  end
end

script.on_event(defines.events.on_player_crafted_item, function(e)
  local player = game.players[e.player_index]
  if e.item_stack.valid_for_read then
    haveitem(player, e.item_stack.name, true)
  end
end)

script.on_event(defines.events.on_picked_up_item, function(e)
  local player = game.players[e.player_index]
  if e.item_stack.valid_for_read then
    haveitem(player, e.item_stack.name, false)
  end
end)

script.on_event(defines.events.on_player_cursor_stack_changed, function(e)
  local player = game.players[e.player_index]
  if player.cursor_stack and player.cursor_stack.valid_for_read then
    haveitem(player, player.cursor_stack.name, false)
  end
end)

script.on_event(defines.events.on_player_main_inventory_changed, function(e)
  local player = game.players[e.player_index]
  local inv = player.get_inventory(defines.inventory.character_main)
  if not inv then -- Compatibility with BlueprintLab_Bud17
    return
  end
  for k,v in pairs(global.unlocks) do
    if not player.force.technologies[v].researched and inv.get_item_count(k) > 0 then
      haveitem(player, k, false)
    end
  end
end)

script.on_init(init)
script.on_configuration_changed(function(cfg)
  init()
  -- Heavy handed fix for mods that forget migration scripts
  for _, force in pairs(game.forces) do
    force.reset_technologies()
    force.reset_recipes()
    for _,tech in pairs(force.technologies) do
      if tech.researched then
        for _, effect in pairs(tech.effects) do
	  if effect.type == "unlock-recipe" then
	    force.recipes[effect.recipe].enabled = true
	  end
	end
      end
    end
    if force.technologies['kovarex-enrichment-process'] then
      force.technologies['kovarex-enrichment-process'].enabled = true
    end
  end
end)

script.on_load(function()
  setpvp()
end)


LandBlock_for_1_1.1.1/prototypes/technology.lua
data:extend({
{
  type = "tool",
  name = "sb-angelsore3-tool",
  localised_name = {"item-name.angels-ore3-crushed"},
  icon = "__angelsrefining__/graphics/icons/angels-ore3-crushed.png",
  icon_size = 32,
  flags = {"hidden"},
  stack_size = 100,
  durability = 1
},
{
  type = "tool",
  name = "sb-basic-circuit-board-tool",
  localised_name = {"item-name.basic-circuit-board"},
  icon = "__bobelectronics__/graphics/icons/basic-circuit-board.png",
  icon_size = 128,
  flags = {"hidden"},
  stack_size = 100,
  durability = 1
},
{
  type = "tool",
  name = "sb-algae-green-tool",
  localised_name = {"item-name.algae-green"},
  icon = "__angelsbioprocessing__/graphics/icons/algae-green.png",
  icon_size = 32,
  flags = {"hidden"},
  stack_size = 100,
  durability = 1
},
{
  type = "tool",
  name = "sb-sulfur-tool",
  localised_name = {"item-name.sulfur"},
  icon = "__angelspetrochem__/graphics/icons/solid-sulfur.png",
  icon_size = 32,
  flags = {"hidden"},
  stack_size = 100,
  durability = 1
},
{
  type = "tool",
  name = "sb-lab-tool",
  localised_name = {"item-name.lab"},
  icon = "__base__/graphics/icons/lab.png",
  icon_size = 64,
  icon_mipmaps = 4,
  flags = {"hidden"},
  stack_size = 100,
  durability = 1
},
{
  type = "technology",
  name = "sb-startup1",
  icon = "__LandBlock_for_1__/graphics/technology/ore.png",
  icon_size = 128,
  effects = {
    {type = "unlock-recipe", recipe = "basic-circuit-board"},
    {type = "unlock-recipe", recipe = "copper-cable"},
    {type = "unlock-recipe", recipe = "algae-farm"},
    {type = "unlock-recipe", recipe = "algae-brown"},
    {type = "unlock-recipe", recipe = "hydro-plant"},
    {type = "unlock-recipe", recipe = "clarifier"},
    {type = "unlock-recipe", recipe = "water-purification"},
    {type = "unlock-recipe", recipe = "iron-chest"},
    {type = "unlock-recipe", recipe = "solid-alginic-acid"}
  },
  unit = {
    count = 1,
    ingredients = {{"sb-angelsore3-tool", 1}},
    time = 5
  }
},
{
  type = "technology",
  name = "sb-startup2",
  icon = "__LandBlock_for_1__/graphics/technology/basic-circuit-board.png",
  icon_size = 128,
  effects = {
    {type = "unlock-recipe", recipe = "inserter"},
    {type = "unlock-recipe", recipe = "algae-green-simple"},
    --{type = "unlock-recipe", recipe = "gas-carbon-dioxide-from-wood"},
    {type = "unlock-recipe", recipe = "pipe"},
    {type = "unlock-recipe", recipe = "pipe-to-ground"},
    {type = "unlock-recipe", recipe = "copper-pipe"},
    {type = "unlock-recipe", recipe = "copper-pipe-to-ground"},
    {type = "unlock-recipe", recipe = "iron-stick"},
    {type = "unlock-recipe", recipe = "iron-gear-wheel"},
    {type = "unlock-recipe", recipe = "burner-inserter"}
  },
  prerequisites = {"sb-startup1"},
  unit = {
    count = 1,
    ingredients = {{"sb-basic-circuit-board-tool", 1}},
    time = 5
  }
},
--[[
{
  type = "technology",
  name = "sb-startup3",
  icon = "__LandBlock_for_1__/graphics/technology/coal.png",
  icon_size = 128,
  effects = {
    {type = "unlock-recipe", recipe = "wood-bricks"},
    {type = "unlock-recipe", recipe = "sb-wood-bricks-charcoal"},
    {type = "unlock-recipe", recipe = "carbon-separation-2"},
    {type = "unlock-recipe", recipe = "lab"}
  },
  prerequisites = {"sb-startup2"},
  unit = {
    count = 1,
    ingredients = {{"sb-algae-green-tool", 1}},
    time = 5
  }
},
{
  type = "technology",
  name = "sb-startup-sulfur",
  icon = "__base__/graphics/technology/sulfur-processing.png",
  icon_size = 128,
  effects = {
    --{type = "unlock-recipe", recipe = "gas-sulfur-dioxide"},
    --{type = "unlock-recipe", recipe = "liquid-sulfuric-acid"},
    --{type = "unlock-recipe", recipe = "yellow-waste-water-purification"}
  },
  prerequisites = {"sb-startup3"},
  unit = {
    count = 1,
    ingredients = {{"sb-sulfur-tool", 1}},
    time = 5
  }
},
]]--
{
  type = "technology",
  name = "sb-startup4",
  icon = "__LandBlock_for_1__/graphics/technology/lab.png",
  icon_size = 128,
  effects = {
    {type = "unlock-recipe", recipe = "automation-science-pack"}
  },
  prerequisites = {"bio-wood-processing"},
  unit = {
    count = 1,
    ingredients = {{"sb-lab-tool", 1}},
    time = 5
  }
}
})

data.raw.recipe['copper-pipe-to-ground'].enabled = false
data.raw.recipe['basic-circuit-board'].enabled = false
if data.raw.recipe['automation-science-pack'].normal then
  data.raw.recipe['automation-science-pack'].normal.enabled = false
  data.raw.recipe['automation-science-pack'].expensive.enabled = false
else
  data.raw.recipe['automation-science-pack'].enabled = false
end
if data.raw.recipe['basic-transport-belt'] then
  table.insert(data.raw.technology['sb-startup2'].effects,
    {type = "unlock-recipe", recipe = "basic-transport-belt"})
else
  table.insert(data.raw.technology['sb-startup2'].effects,
    {type = "unlock-recipe", recipe = "transport-belt"})
end


LandBlock_for_1_1.1.1/prototypes/recipe.lua
data:extend({
  {
    type = "recipe",
    name = "sb-wood-bricks-charcoal",
    category = "smelting",
    enabled = false,
    energy_required = 3.5,
    ingredients = {{"wood-bricks", 1}},
    result = "wood-charcoal",
    result_count = 5,
    subgroup = "petrochem-coal"
  },
  {
    type = "recipe",
    name = "thermal-bore-water",
    category = "thermal-bore",
    subgroup = "water-treatment",
    order = "h[thermal-bore-water]",
    energy_required = 10,
    enabled = false,
    ingredients = {
      {type = "item", name = "lithium-chloride", amount = 1}
    },
    results = {
      {type = "fluid", name = "thermal-water", amount = 5}
    }
  },
  {
    type = "recipe",
    name = "thermal-extractor-water",
    category = "thermal-extractor",
    subgroup = "water-treatment",
    order = "h[thermal-extractor-water]",
    energy_required = 5,
    enabled = false,
    ingredients = {
      {type = "fluid", name = "steam", amount = 100},
      {type = "item", name = "lithium-chloride", amount = 2}
    },
    results = {
      {type = "fluid", name = "thermal-water", amount = 100}
    }
  },
  {
    type = "recipe",
    name = "sb-water-mineralized-crystallization",
    category = "crystallizing",
    subgroup = "slag-processing-1",
    order = "z[slag-processing]",
    enabled = true,
    energy_required = 2,
    ingredients = {
      {type = "fluid", name = "water-mineralized", amount = 200}
    },
    results = {
      {type = "item", name = "angels-ore1", amount = 2, probability = 0.55},
      {type = "item", name = "angels-ore3", amount = 1, probability = 0.7},
    },
    icons = {
      { icon = data.raw.item['angels-ore3'].icon, icon_size = data.raw.item['angels-ore3'].icon_size, scale = 0.5, shift = { 4, 4 } },
      { icon = data.raw.item['angels-ore1'].icon, icon_size = data.raw.item['angels-ore1'].icon_size, scale = 0.5, shift = { -4, -4 }}
    },
    icon_size = 32
  },
  {
    type = "recipe",
    name = "sb-cellulose-foraging",
    localised_name = {"recipe-name.sb-cellulose-foraging"},
    category = "crafting-handonly",
    subgroup = "bio-processing-green",
    enabled = true,
    energy_required = 2,
    ingredients = {},
    results = {
      {type = "item", name = "cellulose-fiber", amount = 1}
    },
    order = "ab[sb-cellulose-foraging]"
  },
  {
    -- Balance assuming blue algae is about equal to green algae in MJ value.
    -- 1 blue cellulose = 2MJ (1 green cellulose = 1MJ but converting to wood pellets doubles it.
    --   Wrong, but I'll stick with it to avoid increasing the cost of all petrochem recipes)
    -- Now for multi phase oil MJ value:
    -- 100 naphtha = 50MJ
    -- 50 fuel oil = 50MJ
    -- basic oil refining is 100 crude oil -> 30 fuel oil + 50 naphtha (and other stuff i'll ignore)
    -- 100 crude oil = 30MJ (fuel oil) + 25MJ (naphtha) = 55MJ
    -- 100 multiphase oil = 55*70/100 = 38.5 MJ.
    -- Let's round that up to 40MJ or 20 blue cellulose fiber.
    -- This is being generous, haven't counted natural gas liquids and base mineral oil.

    -- 100 sulfuric waste water -> 40 blue algae
    -- 40 blue algae -> 20 blue cellulose
    -- 20 blue cellulose -> 100 multi phase oil + 60 sulfuric waste water
    -- 100 multi phase oil -> 10 sulfuric waste water
    -- 70% of sulfuric waste water is recycled
    -- Almost forgot +20 raw gas
    -- 20 raw gas -> 4 acid gas
    -- 4 acid gas -> 2.4 hydrogen sulfide
    -- 2.4 hydrogen sulfide -> 0.12 sulfur
    -- 0.12 sulfur -> 7.2 sulfur dioxide
    -- 7.2 sulfur dioxide -> 4.8 sulfuric acid
    -- 4.8 sulfuric acid -> 16 slag slurry
    -- 16 slag slurry -> 12.8 sulfuric waste water (coal filtering)
    -- So closer to 80% sulfur return
    type = "recipe",
    name = "sb-blue-algae-liquefaction",
    localised_name = {"recipe-name.sb-blue-algae-liquefaction",
      {"fluid-name.liquid-multi-phase-oil"}, {"fluid-name.water-yellow-waste"}, {"fluid-name.gas-carbon-dioxide"}},
    icon_size = 32,
    category = "oil-processing",
    enabled = false,
    energy_required = 5,
    ingredients = {
      {type = "item", name = "blue-cellulose-fiber", amount = 20},
      {type = "fluid", name = "steam", amount = 100}
    },
    results = {
      {type = "fluid", name = "liquid-multi-phase-oil", amount = 100},
      {type = "fluid", name = "water-yellow-waste", amount = 60},
      {type = "fluid", name = "gas-carbon-dioxide", amount = 20}
    },
	main_product = "liquid-multi-phase-oil",
    subgroup = "bio-processing-blue",
    order = "d[blue-algae-liquefaction]"
  }
})

table.insert(data.raw.technology['angels-oil-processing'].prerequisites, 'bio-processing-blue')
table.insert(data.raw.technology['angels-oil-processing'].effects,
  {type = "unlock-recipe", recipe = "sb-blue-algae-liquefaction"})
